#164829209 implement commenting on articles feature

bl4ck4ndbr0wn opened the pull request:
#### Description
As a User:
I want to be able to make a comment(s) on articles, both authored by me and those not authored by me.

#### Type of change
- [ ] Bug fix (nonbreaking change which fixes an issue)
- [x] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [x] End to end
- [x] Integration

#### Checklist:
- [x] My code follows the style guide for this project
- [x] I have linted my code prior to submission
- [ ]  I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### How should this be manually tested?
1.  Clone the repository 
```https://github.com/andela/ah-centauri-backend.git```
2. On completing of clonning
```bash 
cd ah-centauti-backend
git checkout feature/164829209-user-can-comment-on-article
```
3. Rename the .env.example file into .env and update the parameters with your own keys.

> For Centauri team members I will share this keys through slack

4. Start development server and send the following through postman

**Displaying all comments for an Article**
```
GET /api/articles/:slug/comments
returns All Comment. 
```
**Displaying all comments replies for an Article**
```
GET /api/articles/:slug/comments?parent=9
returns All Comment. 
```
**Add Comments to an Article**
```
POST /api/articles/:slug/comments

Example request body:

{
  "comment": {
    "body": "His name was my name too."
  }
}
Authentication required returns the created Comment. Required field: body
```
**Add a Reply to a Comment in an Article**
```
POST /api/articles/:slug/comments

Example request body:

{
  "comment": {
    "body": "His name was my name too.",
    "parent": 9
  }
}

args: 
       :parent: is the ID of the comment you are repling to.

Authentication required returns the created Comment. Required field: body
```
**Update a Comment to an Article**
```
PATCH /api/articles/:slug/comments/:id

Example request body:

{
  "comment": {
    "body": "His name was my name too."
  }
}
Authentication required, returns the created Comment. Required field: body
```
**Delete Comment**
```
DELETE /api/articles/:slug/comments/:id

Authentication required.
```

#### Pivotal Tracker
[#164829209](https://www.pivotaltracker.com/n/projects/2321088/stories/164829209)

Sm0k3s commented:
Looks pragmatic to me

ValerieRono reviewed (changes requested):
1). change urls to follow conventions
2). replies to a comment's comment not reflecting

bl4ck4ndbr0wn committed:
164829208(article-comments)-implement-comments-on-articles
- Resolved pep8 errors in the entire project files
- Only authenticated users should be able to comments on articles.
- Successive comments should be threaded
- Timestamped comments.
- write tests for implemented feature

[Delivers #164829208]": Users should be able to comment on articles"

oagutu commented on authors/apps/comments/tests/test_comment_view.py:
```diff
@@ -0,0 +1,322 @@
+def specific_comment_url(slug, comment_id):
+    return reverse('comments:list_comment', args=[slug, comment_id])
+
+
```
Why not have a `BaseTestCase` with a "parent"  `setUp` method. The Test classes in this file can then simply inherit form the `BaseTestCase`?

oagutu commented on authors/apps/comments/views.py:
```diff
@@ -0,0 +1,147 @@
+        except Comment.DoesNotExist:
+            return Response({"errors": "Specified Comment does not exist."}, status=status.HTTP_404_NOT_FOUND)
+
+    def post(self, request, *args, **kwargs):
```
Can't this already be handled by the `post` method in the class ` RetrieveCommentAPIView` above?

bl4ck4ndbr0wn replied:
The post in:
RetrieveCommentAPIView requires one argument which is the slug to create a new comment.
RetrieveUpdateDeleteCommentAPIView requires two arguments which are the slug and comment id to create a new comment[reply]

bl4ck4ndbr0wn replied:
Each view has it own url that limits what can be sent by the user.

oagutu commented on authors/apps/comments/views.py:
```diff
@@ -0,0 +1,147 @@
+            comment = Comment.objects.get(pk=self.kwargs['id'])
+            self.check_object_permissions(request, comment)
+            comment.delete()
+            return Response({'message': "Deletion successful."}, status.HTTP_204_NO_CONTENT)
```
`204` responses should not contain a message body

bl4ck4ndbr0wn replied:
> @bl4ck4ndbr0wn good job with this. Look into the issues raised, including the following:
> Improve test coverage
> `GET` - `/api/articles/<slug>/comments/`:
> 
> * Repetition of comments in response ie. A child is returned under a parent as well as on its own

In regard to test coverage currently, I have tested everything inside the comments application.
I have uploaded an image showing the coverage in my feature.
![image](https://user-images.githubusercontent.com/33542127/55937456-b329c700-5c41-11e9-9a36-b8ebedefa569.png)

Maybe to add the coverage if it is ok, i can test other apps in the project folder.

oagutu reviewed (commented):
@bl4ck4ndbr0wn good job with this. Look into the issues raised, including the following:
Improve test coverage
`GET` - `/api/articles/<slug>/comments/`: 
- Repetition of comments in response ie. A child is returned under a parent as well as on its own

ogiste commented on authors/apps/comments/views.py:
```diff
@@ -0,0 +1,147 @@
+            return Response(serializer.data,
+                            status=status.HTTP_200_OK)
+        except Articles.DoesNotExist:
+            return Response({"errors": "Specified Article does not exist."}, status=status.HTTP_404_NOT_FOUND)
```
Create a separate file for the response messages sent to in a response, makes it easier to update.
